packages/owl-bot/src/database.ts

```diff
@@ -0,0 +1,138 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import admin from 'firebase-admin';
+import {OwlBotLock} from './config-files';
+import {Configs, ConfigsStore} from './configs-store';
+
+export type Db = admin.firestore.Firestore;
+interface UpdateLockPr {
+  pullRequestId: string;
+}
+
+/**
+ * When firebase sees a / in a doc id, it interprets it as a collection name.
+ * So, we have to escape them.  Also escape +s because we use them for combining
+ * strings into keys.  And use a format that can be decoded by decodeURIComponent().
+ */
+export function encodeId(s: string): string {
+  return s.replace('%', '%25').replace('/', '%2F').replace('+', '%2B');
+}
+
+export function decodeId(s: string): string {
+  return decodeURIComponent(s);
+}
+
+function makeUpdateLockKey(repo: string, lock: OwlBotLock): string {
+  return [repo, lock.docker.image, lock.docker.digest].map(encodeId).join('+');
+}
+
+export class FirestoreConfigsStore implements ConfigsStore {
+  private db: Db;
+  readonly yamls: string;
+  readonly lock_update_prs: string;
+
+  /**
+   * @param collectionsPrefix should only be overridden in tests.
+   */
+  constructor(db: Db, collectionsPrefix = 'owl-bot-') {
```

Comment:
I think we need to pass in a project name somehow, since our Firestore database is on a different project.

Perhaps you could read from `process.env.FIREBASE_PROJECT`?

Reply:
Talked out of band, this will be done in a follow up PR.